sycl/doc/extensions/proposed/sycl_ext_oneapi_peer_access.asciidoc

```diff
@@ -0,0 +1,158 @@
+access to a peer device's memory inside a compute kernel and optimized memory
+copies between peer devices.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_PEER_ACCESS` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== Peer to Peer (P2P) Memory Access APIs
+
+This extension adds support for mechanisms to query and enable support for
+direct memory access between peer devices in a system.
+In particular, this allows one device to directly access USM Device
+allocations for a peer device in the same context.
+Peer to peer capabilities are useful as they can provide access to a peer
+device's memory inside a compute kernel and also optimized memory copies between
+peer devices.
+
+This extension adds the following new member functions to the device class, as described
+below.
+
+[source,c++]
+----
+namespace sycl   {
+namespace ext    {
+namespace oneapi {
+enum class peer_access {
+  access_supported,
+  access_enabled,
+  atomics_supported,
+};
+} // namespace oneapi
+} // namespace ext
+
+class device {
+public:
+  bool ext_oneapi_can_access_peer(const device &peer,
+                                  ext::oneapi::peer_access value =
+                                    ext::oneapi::peer_access::access_supported);
+  void ext_oneapi_enable_peer_access(const device &peer);
+  void ext_oneapi_disable_peer_access(const device &peer);
+};
+
+} // namespace sycl
+----
+
+The semantics of the new functions are:
+
+|===
+|Member Function |Description
+
+|bool ext_oneapi_can_access_peer(const device &peer,
+                                  ext::oneapi::peer_access value =
+                                  ext::oneapi::peer_access::access_supported)
+a|Queries the peer access status between this device and `peer` according to
+the query `value`:
+
+* `ext::oneapi::peer_access::access_supported`: Returns true only if it is
+possible for this device to enable peer access to USM device memory allocations
+located on the `peer` device.
+
+* `ext::oneapi::peer_access::access_enabled`: Returns true only if peer access is
+currently enabled from this device to the `peer` device.
+
+* `ext::oneapi::peer_access::atomics_supported`: When this query returns true,
+it indicates that this device may perform atomic operationson USM device memory
+allocations located on the `peer` device when peer access is enabled to that
+device. If the query returns false, attempting to perform atomic operations on
+`peer` memory will have undefined behavior.
```

Comment:
The core SYCL spec makes a distinction between "atomic operations" and "concurrent access".  The Level Zero driver has separate  queries for these two concepts.  We need to clarify what `atomics_supported` means.  I think it should mean that both atomic operations and concurrent access is supported, which is consistent with the current wording in the SYCL spec for the `usm_atomic_shared_allocations` aspect.

This is an area we are debating in general, though, so we may end up making two different queries for these concepts.

Reply:
I don't think concurrent access comes into play here - I think it's only (pseudocode) atomicAdd(ptr, val)

Reply:
Atomic operations only make sense if two things can access the memory concurrently.  I guess there are two possible interpretations for what `atomics_supported` means:

1. This device and `peer` device can concurrently access the device USM and do atomic operations on that memory.  These operations are atomic w.r.t. code running on the two devices.

2. This device can access device USM from `peer`, but it cannot access it concurrently with `peer`.  Atomic operations are supported, but only between work-items running on this device.

I was originally thinking the query meant (1), but your comment makes me think that maybe you intend (2)?

Reply:
Another thing that we should pay attention to here is the concept of [memory scope](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_memory_scope).

If the device and `peer` can use atomics to concurrently update the same memory, then both devices will need to list `memory_scope::system` in `info::device::atomic_memory_scope_capabilities`.   Both devices will need to use atomics with `memory_scope::system` when concurrently accessing the memory to avoid a data race.

If the device is only accessing `peer`'s memory atomically but not concurrently with `peer`, it can use atomics with `memory_scope::device`. If `peer` accesses the same memory concurrently, that's a data race.

I don't know whether it's better to use the atomics & concurrent distinction or to work in some concept of scope, but I agree with Greg that this needs to be clarify exactly what is guaranteed.